cranelift/interpreter/src/interpreter.rs

```diff
@@ -0,0 +1,362 @@
+        }
+    }
+}
+
+/// The ways interpretation can fail.
+#[derive(Error, Debug)]
+pub enum Trap {
+    #[error("unknown trap")]
+    Unknown,
+    #[error("invalid type for {1}: expected {0}")]
+    InvalidType(String, ValueRef),
+    #[error("invalid cast")]
+    InvalidCast(#[from] DataValueCastFailure),
+    #[error("the instruction is not implemented (perhaps for the given types): {0}")]
+    Unsupported(Inst),
+    #[error("reached an unreachable statement")]
+    Unreachable,
+    #[error("invalid control flow: {0}")]
+    InvalidControlFlow(String),
+    #[error("invalid function reference: {0}")]
+    InvalidFunctionReference(FuncRef),
+    #[error("invalid function name: {0}")]
+    InvalidFunctionName(String),
+}
+
+/// The Cranelift interpreter; it contains immutable elements such as the function environment and
+/// implements the Cranelift IR semantics.
+#[derive(Default)]
+pub struct Interpreter {
+    pub env: Environment,
+}
+
+/// Helper for more concise matching.
+macro_rules! binary_op {
+    ( $op:path[$arg1:ident, $arg2:ident]; [ $( $data_value_ty:ident ),* ]; $inst:ident ) => {
+        match ($arg1, $arg2) {
+            $( (DataValue::$data_value_ty(a), DataValue::$data_value_ty(b)) => { Ok(DataValue::$data_value_ty($op(a, b))) } )*
+            _ => Err(Trap::Unsupported($inst)),
+        }
+    };
+}
+
+impl Interpreter {
+    /// Construct a new [Interpreter] using the given [Environment].
+    pub fn new(env: Environment) -> Self {
+        Self { env }
+    }
+
+    /// Call a function by name; this is a helpful proxy for [Interpreter::call_by_index].
+    pub fn call_by_name(
+        &self,
+        func_name: &str,
+        arguments: &[DataValue],
+    ) -> Result<ControlFlow, Trap> {
+        let func_ref = self
+            .env
+            .index_of(func_name)
+            .ok_or_else(|| Trap::InvalidFunctionName(func_name.to_string()))?;
+        self.call_by_index(func_ref, arguments)
+    }
+
+    /// Call a function by its index in the [Environment]; this is a proxy for [Interpreter::call].
+    pub fn call_by_index(
+        &self,
+        func_ref: FuncRef,
+        arguments: &[DataValue],
+    ) -> Result<ControlFlow, Trap> {
+        match self.env.get_by_func_ref(func_ref) {
+            None => Err(Trap::InvalidFunctionReference(func_ref)),
+            Some(func) => self.call(func, arguments),
+        }
+    }
+
+    /// Interpret a call to a [Function] given its [DataValue] arguments.
+    fn call(&self, function: &Function, arguments: &[DataValue]) -> Result<ControlFlow, Trap> {
+        debug!("Call: {}({:?})", function.name, arguments);
+        let first_block = function
+            .layout
+            .blocks()
+            .next()
+            .expect("to have a first block");
+        let parameters = function.dfg.block_params(first_block);
+        let mut frame = Frame::new(function).with_parameters(parameters, arguments);
```

Comment:
Check that `arguments.len() == parameters.len()`?

Reply:
The assertion is inside `Frame::with_parameters`